File: Java_MVP/app/src/main/java/com/gotoinc/requesin/repository/ApiContractImpl.java
package com.gotoinc.requesin.repository;

import android.util.Log;

import com.gotoinc.requesin.view.BaseViewContract;
import com.gotoinc.requesin.view._model.User;
import com.gotoinc.requesin.view.home.HomeContract;

import java.io.IOException;
import java.net.SocketTimeoutException;

import androidx.annotation.NonNull;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import retrofit2.HttpException;

/**
 * Created by Illia Derevianko on 23.11.18.
 * GoTo Inc.
 */
public class ApiContractImpl implements ApiContract {
    private RetrofitRequest queries;

    public ApiContractImpl(RetrofitRequest queries) {
        this.queries = queries;
    }

    @Override
    public void getUsersList(int page, @NonNull HomeContract view, @NonNull BaseViewContract baseView) {
        queries.getUsersList(page)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(response -> {
                        if(response.body() != null) {
                            view.setUsers(User.from(response.body().getData()));
                            if(response.body().getTotalPages() > response.body().getPage()) {
                                view.nextPage(true, response.body().getPage() + 1);
                            } else {
                                view.nextPage(false, page);
                            }
                        }
                    }
                , e -> {
                        if (e instanceof HttpException) {
                            baseView.onError();
                        } else if (e instanceof SocketTimeoutException) {
                            baseView.onTimeout();
                        } else if (e instanceof IOException) {
                            baseView.onNetworkError();
                        } else {
                            baseView.onError(e.getMessage());
                        }
                    });
    }
}

File: Kotlin_MVP/app/src/main/java/com/gotoinc/requesin/repository/ApiContractImpl.kt
package com.gotoinc.requesin.repository

import com.gotoinc.requesin.repository.datamodel.PaginationResponse
import com.gotoinc.requesin.repository.datamodel.UserResponse
import com.gotoinc.requesin.view.BaseViewContract
import com.gotoinc.requesin.view.home.HomeContract
import com.gotoinc.requesin.view.model.User
import retrofit2.Call
import retrofit2.Callback
import retrofit2.HttpException
import retrofit2.Response
import java.io.IOException
import java.net.SocketTimeoutException

/**
 * Created by Illia Derevianko on 28.11.18.
 * GoTo Inc.
 */
class ApiContractImpl(private val queries: RetrofitRequest) : ApiContract {
    override fun getUsersList(page: Int, view: HomeContract, baseView: BaseViewContract) {
        queries.getUserList(page)
            .enqueue(object: Callback<PaginationResponse<List<UserResponse>>> {
                override fun onResponse(
                    call: Call<PaginationResponse<List<UserResponse>>>,
                    response: Response<PaginationResponse<List<UserResponse>>>
                ) {
                    val paginationResponse= response.body()
                    paginationResponse?.let {
                        view.setUsers(User.from(paginationResponse.data))
                        val nextPage = paginationResponse.page + 1
                        if (paginationResponse.totalPages >= nextPage)
                            view.nextPage(true, nextPage)
                        else
                            view.nextPage(false, paginationResponse.page)
                    }
                }

                override fun onFailure(call: Call<PaginationResponse<List<UserResponse>>>, t: Throwable) {
                    when (t) {
                        is HttpException -> baseView.onError()
                        is SocketTimeoutException -> baseView.onTimeout()
                        is IOException -> baseView.onNetworkError()
                        else -> baseView.onError(t.localizedMessage)
                }
            }
        })
    }

}